test/inductor/test_pattern_matcher.py

```diff
@@ -1655,6 +1657,78 @@ def my_func_static(x, w, epsilon):
         test, (code,) = run_and_get_code(my_func_static, *inputs)
         self.assertTrue("static_scaled_int8_quant" not in code)
 
+    def test_mutable_op_nonview_inputs_register_replacement(self):
+        @torch.library.custom_op("mylib::foo_inplace", mutates_args={"x"})
+        def foo_inplace(x: torch.Tensor) -> None:
+            x.add_(1)
+
+        # NOTE: only returning None is supported; the custom op cannot return `out` because it's part of op input.
+        @torch.library.custom_op("mylib::bar", mutates_args={"out"})
+        def bar_out(x: torch.Tensor, out: torch.Tensor) -> None:
+            out.copy_(x + 2)
+
+        @register_fake("mylib::bar")
+        def bar_out_fake(x: torch.Tensor, out: torch.Tensor) -> None:
+            return None
+
+        @torch.library.custom_op("mylib::foobar_out", mutates_args={"out"})
+        def foobar_out(x: torch.Tensor, out: torch.Tensor) -> None:
+            x.add_(1)
+            out.copy_(x + 7)  # intentionally different from bar_out
+
+        def mutable_ops_pattern(x, out):
+            foo_inplace(x)
+            bar_out(x, out)
+            return out
+
+        def mutable_ops_replacement(x, out):
+            foobar_out(x, out)
+            return out
+
+        inp = torch.randn(3)
+
+        my_patterns = PatternMatcherPass()
+        register_replacement(
+            search_fn=mutable_ops_pattern,
+            replace_fn=mutable_ops_replacement,
+            example_inputs=[inp.clone().detach(), inp.clone().detach()],
+            trace_fn=functools.partial(fwd_only, apply_auto_functionalize=True),
```

Comment:
1) Should apply_auto_functionalize be the default for fwd_only? Since Inductor pattern matching happens only at the post-grad pass?
2) If not, should we rename `apply_auto_functionalize` something like `post_grad`? Because that is where we plan to use the pattern matcher